rasa/utils/tensorflow/rasa_layers.py

```diff
@@ -0,0 +1,940 @@
+            one for each feature type (`sequence` or `sentence`) of the given attribute.
+        config: A model config used for correctly parameterising the layer and the
+            `ConcatenateSparseDenseFeatures` layer it uses internally.
+
+    Input shape:
+        Tuple of three input tensors:
+            sequence_features: List of 3-D dense or sparse tensors, each with shape
+                `(batch_size, max_seq_length, input_dim)` where `input_dim` can be
+                different for sparse vs dense tensors. See the input shape of
+                `ConcatenateSparseDenseFeatures` for more information.
+            sentence_features: List of 3-D dense or sparse tensors, each with shape
+                `(batch_size, 1, input_dim)` where `input_dim` can be different for
+                sparse vs dense tensors, and can differ from that in
+                `sequence_features`. See the input shape of
+                `ConcatenateSparseDenseFeatures` for more information.
+            sequence_feature_lengths: Dense tensor of shape `(batch_size, )`.
+
+    Output shape:
+        combined_features: A 3-D tensor with shape `(batch_size, sequence_length,
+            units)` where `units` is  completely  determined by the internally applied
+            `ConcatenateSparseDenseFeatures` layer and `sequence_length` is the combined
+            length of sequence- and sentence-level features: `max_seq_length + 1` if
+            both feature types are present, `max_seq_length` if only sequence-level
+            features are present, and 1 if only sentence-level features are present).
+        mask_combined_sequence_sentence: A 3-D tensor with shape
+            `(batch_size, sequence_length, 1)`.
+
+    Raises:
+        A `TFLayerConfigException` if no feature signatures are provided.
+
+    Attributes:
+        output_units: The last dimension size of the layer's `combined_features` output.
+    """
+
+    def __init__(
+        self,
+        attribute: Text,
+        attribute_signature: Dict[Text, List[FeatureSignature]],
+        config: Dict[Text, Any],
+    ) -> None:
+        if not attribute_signature or not (
+            attribute_signature.get(SENTENCE, [])
+            or attribute_signature.get(SEQUENCE, [])
+        ):
+            raise TFLayerConfigException(
+                "The attribute signature must contain some feature signatures."
+            )
+
+        super().__init__(name=f"rasa_feature_combining_layer_{attribute}")
+
+        self._tf_layers = {}
+
+        # Prepare sparse-dense combining layers for each present feature type
+        self._feature_types_present = self._get_present_feature_types(
+            attribute_signature
+        )
+        self._prepare_sparse_dense_concat_layers(attribute, attribute_signature, config)
+
+        # Prepare components for combining sequence- and sentence-level features
+        self.output_units = self._prepare_sequence_sentence_concat(attribute, config)
```

Comment:
@twerkmeister this is the issue. `_prepare_sequence_sentence_concat` outputs `output_units` as a by-product, which isn't reflected in the method's name. But the logic for creating sequence-sentence layers and for determining `output_units` (inside `_prepare_sequence_sentence_concat`) is so intertwined that I think I'd literally have do duplicate most of it if I were to have two methods (called e.g. `_prepare_sequence_sentence_concat` and `_calculate_output_units`). At the same time, I think `_prepare_sequence_sentence_concat` describes the most important thing that the method does, so I'm hesitant to rename it...

Reply:
Yeah it is indeed somewhat intertwined. One possibility for splitting it up could be something like the following:
1. have a method `maybe_create_unifying_layers` which creates the unifying layers if necessary but does nothing else.
2. have a method `determine_own_output_units` :
```
if some unifying layers present:
    return config concat dim
elif feature present SEQUENCE:
    return sequence out dim
else 
    return sentence out dim
```
It's a bit of a shortcut, but I think the logic is right. If the unifying layers are present, obviously we take the concat dim. If they are not present, that means either: 1) both feature types are present and have the same dim, or 2) only one feature is present. In either case it is fine to return the output units of the first feature we find present. 

What do you think?

Reply:
The downside is that `determine_own_output_units` should only be called after `maybe_create_unifying_layers` otherwise the logic won't work out. So maybe just leave it as it is, or make the logic safer haha. Maybe you can put some of the possibly duplicate case logic between the two methods in yet another function.

Reply:
Thanks @twerkmeister! I see that there's no ideal solution. I decided to split the method into 2 which have to be called in a strict order.